Source/Core/Common/Network.cpp

```diff
@@ -186,4 +261,204 @@ u16 ComputeNetworkChecksum(const void* data, u16 length, u32 initial_value)
+                                &result[crc_offset], (u16)result.size() - crc_offset, IPPROTO_TCP);
+  std::memcpy(&result[crc_offset], &tcp_header, TCPHeader::SIZE);
+  return result;
+}
+
+u16 TCPPacket::Size() const
+{
+  return (u16)(MIN_SIZE + data.size() + ipv4_options.size() + tcp_options.size());
+}
+
+UDPPacket::UDPPacket() = default;
+
+UDPPacket::UDPPacket(MACAddress destination, MACAddress source)
+{
+  eth_header.destination = destination;
+  eth_header.source = source;
+  eth_header.ethertype = htons(IPV4_ETHERTYPE);
+}
+
+std::vector<u8> UDPPacket::Build()
+{
+  std::vector<u8> result;
+  result.resize(Size());
+
+  // recalc size
+  ip_header.total_len =
+      htons((u16)(IPv4Header::SIZE + ipv4_options.size() + UDPHeader::SIZE + data.size()));
+  udp_header.length = htons((u16)(UDPHeader::SIZE + data.size()));
+
+  // copy data
+  std::memcpy(result.data(), &eth_header, EthernetHeader::SIZE);
+  std::memcpy(&result[EthernetHeader::SIZE], &ip_header, IPv4Header::SIZE);
+  u16 offset = EthernetHeader::SIZE + IPv4Header::SIZE;
+  if (ipv4_options.size() > 0)
+  {
+    std::memcpy(&result[offset], ipv4_options.data(), ipv4_options.size());
+    offset += (u16)ipv4_options.size();
+  }
+  std::memcpy(&result[offset], &udp_header, UDPHeader::SIZE);
+  const u16 crc_offset = offset;
+  offset += UDPHeader::SIZE;
+  if (data.size() > 0)
+  {
+    std::memcpy(&result[offset], data.data(), data.size());
+  }
+  udp_header.checksum = 0;
+  udp_header.checksum =
+      ComputeTCPNetworkChecksum(ip_header.source_addr, ip_header.destination_addr,
+                                &result[crc_offset], (u16)result.size() - crc_offset, IPPROTO_UDP);
+  std::memcpy(&result[crc_offset], &udp_header, UDPHeader::SIZE);
+
+  return result;
+}
+
+u16 UDPPacket::Size() const
+{
+  return (u16)(MIN_SIZE + data.size() + ipv4_options.size());
+}
+
+PacketView::PacketView(const u8* ptr, std::size_t size) : m_ptr(ptr), m_size(size)
+{
+}
+
+std::optional<u16> PacketView::GetEtherType() const
+{
+  if (m_size < EthernetHeader::SIZE)
+    return std::nullopt;
+  const std::size_t offset = offsetof(EthernetHeader, ethertype);
+  return ntohs(Common::BitCastPtr<u16>(m_ptr + offset));
+}
+
+std::optional<ARPPacket> PacketView::GetARPPacket() const
+{
+  if (m_size < ARPPacket::SIZE)
+    return std::nullopt;
+  return Common::BitCastPtr<ARPPacket>(m_ptr);
+}
+
+std::optional<u8> PacketView::GetIPProto() const
+{
+  if (m_size < EthernetHeader::SIZE + IPv4Header::SIZE)
+    return std::nullopt;
+  return m_ptr[EthernetHeader::SIZE + offsetof(IPv4Header, protocol)];
+}
+
+std::optional<TCPPacket> PacketView::GetTCPPacket() const
+{
+  if (m_size < TCPPacket::MIN_SIZE)
+    return std::nullopt;
+  TCPPacket result;
+  result.eth_header = Common::BitCastPtr<EthernetHeader>(m_ptr);
+  result.ip_header = Common::BitCastPtr<IPv4Header>(m_ptr + EthernetHeader::SIZE);
+  const u16 offset = ((result.ip_header.version_ihl & 0xf) * 4) + EthernetHeader::SIZE;
```

Comment:
These expressions `ex: (ip_header.version_ihl & 0xf) * 4` are used quite a lot and should have their own method. It will make it less redundant and less likely to make mistakes.